### PR TITLE
Replace throttler algorithm by token bucket

### PR DESCRIPTION
**Pull request by serxa**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Throttling algorithm changed to token bucket

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

File: src/Common/Throttler.cpp [hunk 1]
```diff
@@ -21,63 +20,56 @@ namespace ErrorCodes
...
-            /// We want to make old values of speed less important for our smoothed value
-            /// so we decay it's value with coef.
-            auto decay_coeff = std::pow(0.5, time_spent / window_ns);
-
-            /// Weighted average between previous and new speed
-            smoothed_speed = smoothed_speed * decay_coeff + (1 - decay_coeff) * new_speed;
-            current_speed = smoothed_speed;
+            double delta_seconds = static_cast<double>(now - prev_ns) / NS;
+            tokens = std::min<double>(tokens + max_speed * delta_seconds - amount, max_burst);
         }
```
Comment (yakov-olkhovskiy) [inline src/Common/Throttler.cpp:55]: I'm not sure about first call to `add` - there will no throttling at all - maybe better approach will be something like this:
```
        if (max_speed)
        {
            double delta_seconds = prev_ns ? static_cast<double>(now - prev_ns) / NS : 0;
            tokens = std::min<double>(tokens + max_speed * delta_seconds - amount, max_burst);
        }
```

File: src/Common/Throttler.h [hunk 2]
```diff
@@ -10,25 +10,26 @@
...
+  * Implemented using Token Bucket Throttling algorithm.
+  * Also allows you to set a limit on the maximum number of tokens. If exceeded, an exception will be thrown.
   */
 class Throttler
 {
 public:
-    explicit Throttler(size_t max_speed_, const std::shared_ptr<Throttler> & parent_ = nullptr)
-            : max_speed(max_speed_), limit_exceeded_exception_message(""), parent(parent_) {}
+    Throttler(size_t max_speed_, size_t max_burst_, const std::shared_ptr<Throttler> & parent_ = nullptr)
+        : max_speed(max_speed_), max_burst(max_burst_), limit_exceeded_exception_message(""), tokens(max_burst), parent(parent_) {}
```
Comment (yakov-olkhovskiy) [inline src/Common/Throttler.h:21]: Shouldn't we limit `max_burst` somehow? by `max_speed` for example...

File: src/Common/Throttler.cpp [hunk 3]
```diff
@@ -21,63 +20,56 @@ namespace ErrorCodes
...
-        {
-            accumulated_sleep += sleep_time;
-
-            sleepForNanoseconds(sleep_time);
-
-            accumulated_sleep -= sleep_time;
-
-            ProfileEvents::increment(ProfileEvents::ThrottlerSleepMicroseconds, sleep_time / 1000UL);
-        }
+        int64_t sleep_time = static_cast<int64_t>(-tokens_value * max_speed * NS);
```
Comment (yakov-olkhovskiy) [inline src/Common/Throttler.cpp:68]: this seems wrong to me - dimensional analysis doesn't add up - tokens * tokens/sec = tokens^2/sec - it should end up in seconds - probably this is correct:
```suggestion
        int64_t sleep_time = static_cast<int64_t>(-tokens_value / max_speed * NS);
```

Comment (serxa) [reply src/Common/Throttler.cpp:68]: Right, thanks!

File: src/Common/Throttler.h (hunk 2 above)
Comment (serxa) [reply src/Common/Throttler.h:21]: No, it is independent, zero and very large values also can make sense in some scenarios. I'm going to add them in config in separate PR

**Review by yakov-olkhovskiy — APPROVED** [review]

**Comment by serxa** [comment]
Simple manual tests show that everything is working properly:
```
:) SET max_network_bandwidth = 100000;

Ok.

:) SELECT sum(ignore(*)) FROM (SELECT rand64() FROM remote('127.0.0.{2,3}', numbers(2000000)));

1 row in set. Elapsed: 320.717 sec. Processed 4.06 million rows, 32.44 MB (12.64 thousand rows/s., 101.16 KB/s.)
```

**Pull request merged**